test/PushNotificationTests/BaseTestSuite.cpp

```diff
@@ -0,0 +1,264 @@
+    if (channelOperation.wait_for(c_timeout) != winrt::Windows::Foundation::AsyncStatus::Completed)
+    {
+        channelOperation.Cancel();
+        VERIFY_FAIL(L"Channel request hit timeout.");
+    }
+
+    auto result{ channelOperation.GetResults() };
+    VERIFY_ARE_EQUAL(result.Status(), PushNotificationChannelStatus::CompletedSuccess);
+
+    result.Channel().Close();
+    VERIFY_THROWS_HR(result.Channel().Close(), WPN_E_CHANNEL_CLOSED);
+}
+
+void BaseTestSuite::VerifyRegisterAndUnregister()
+{
+    PushNotificationManager::Default().Register();
+    m_registered = true;
+
+    PushNotificationManager::Default().Unregister();
+}
+
+void BaseTestSuite::VerifyUnregisterFails()
+{
+    if (PushNotificationManager::Default().IsSupported())
+    {
+        VERIFY_THROWS_HR(PushNotificationManager::Default().Unregister(), E_UNEXPECTED);
+    }
+    else
+    {
+        PushNotificationManager::Default().Unregister();
+    }
+}
+
+void BaseTestSuite::VerifyUnregisterAllFails()
+{
+    if (PushNotificationManager::Default().IsSupported())
+    {
+        VERIFY_THROWS_HR(PushNotificationManager::Default().UnregisterAll(), E_UNEXPECTED);
+    }
+    else
+    {
+        PushNotificationManager::Default().UnregisterAll();
+    }
+}
+
+void BaseTestSuite::VerifyRegisterAndUnregisterAll()
+{
+    PushNotificationManager::Default().Register();
+    m_registered = true;
+
+    PushNotificationManager::Default().UnregisterAll();
+    m_registered = false;
+}
+
+void BaseTestSuite::MultipleRegister()
+{
+    PushNotificationManager::Default().Register();
+    m_registered = true;
+
+    if (PushNotificationManager::Default().IsSupported())
+    {
+        VERIFY_THROWS_HR(PushNotificationManager::Default().Register(), HRESULT_FROM_WIN32(ERROR_ALREADY_REGISTERED));
+    }
+    else
+    {
+        PushNotificationManager::Default().Register();
+    }
+}
+
+void BaseTestSuite::VerifyMultipleRegisterAndUnregister()
+{
+    PushNotificationManager::Default().Register();
+    m_registered = true;
+
+    PushNotificationManager::Default().Unregister();
+
+    if (PushNotificationManager::Default().IsSupported())
+    {
+        VERIFY_THROWS_HR(PushNotificationManager::Default().Register(), HRESULT_FROM_WIN32(ERROR_ALREADY_REGISTERED));
+    }
+    else
+    {
+        PushNotificationManager::Default().Register();
+    }
+}
+
+void BaseTestSuite::VerifyMultipleRegisterAndUnregisterAll()
+{
+    PushNotificationManager::Default().Register();
+    m_registered = true;
+
+    PushNotificationManager::Default().UnregisterAll();
+    m_registered = false;
+
+    PushNotificationManager::Default().Register();
+    m_registered = true;
+
+    PushNotificationManager::Default().UnregisterAll();
+    m_registered = false;
+}
+
+void BaseTestSuite::VerifyUnregisterTwice()
+{
+    PushNotificationManager::Default().Register();
+    m_registered = true;
+
+    PushNotificationManager::Default().Unregister();
```

Comment:
Shouldn't m_registered be set to false after this? Needs a helper function as pointed to above feedback.

Reply:
Unregister doesn't fully unregister an app from PushNotifications. UnregisterAll wipes the states completely clean.

Reply:
Then we need a tri-state and a bool won't suffice in the tests. Tri state could be mapped by an enum:
enum RegistrationState{
  Registered;
  UnregisteredPartial;
  UnREgisteredFull;
}

Default state at the beginning of every test would be UnregisteredFull. We would then move from UnregisteredFull->Registered
If Unregister is called, we would move from Registered->UnregisteredPartial;
If Unregisterall is called, we would move to UnregisteredFull;

Reply:
Do we need the middle state? At least in terms of test environment, I don't think it's needed.

Reply:
You could get away without it, but the test logic is a bit convoluted without it. Unregister also means the bool is false. Maybe call it m_UnregisteredFully?